### PR TITLE
[Workflow] Commenting

### PR DESCRIPTION
**Pull request by jacobtoppm**
- Adds the ability for workflow tasks/actions to provide modals with formats for extra data entry
- This is initially used to add comments on approve/reject actions
- Adds comment to the rejection email
- Makes the needs-changes state banner orange and reorders the workflow timeline to show most recent tasks/comments first, following demo feedback

### CODE REVIEW

File: wagtail/admin/mail.py [hunk 1]
```diff
@@ -328,11 +328,14 @@ def get_recipient_users(self, workflow_state, **kwargs):
...
 
+    task_classes = {GroupApprovalTask, }
+
     def __init__(self):
         super().__init__((TaskState,))
 
     def can_handle(self, instance, **kwargs):
-        if super().can_handle(instance, **kwargs) and isinstance(instance.task.specific, GroupApprovalTask):
+        if super().can_handle(instance, **kwargs) and instance.task.specific.__class__ in self.task_classes:
+            # do not use isinstance so subclasses of GroupApprovalTask can define their own Notifiers if wanted
```
Comment (gasman) [inline wagtail/admin/mail.py:346]: I'm in two minds about this - this means that if you subclass GroupApprovalTask for something really minor or unrelated to notifications, then you lose the notification functionality until you connect up a new signal handler for that subclass. This seems to go against the spirit of what object inheritance is for, and could be as much of a 'gotcha' as the double emails.

I'm not really sure how to fix that, though - it probably calls for a mechanism where each task subclass is only associated with a single notifier. Given that this change seems to be a bit tangential to the rest of this PR, would it be OK to leave this out for now and work out a better solution later?

File: wagtail/admin/views/pages.py [hunk 2]
```diff
@@ -1279,25 +1286,63 @@ def workflow_action(request, page_id):
...
+                    json_data={'step': 'action'}
+                )
+        else:
+            response = task.on_action(task_state, request.user, action_name)
+        if response:
+            return response
+        return redirect(redirect_to)
+    else:
+        form = form_class()
+    return render_modal_workflow(
```
[the remark this answers is not in the document]
Comment (jacobtoppm) [reply wagtail/admin/views/pages.py]: Thanks, good point. Done now!

File: wagtail/admin/templates/wagtailadmin/pages/workflow_action_modal.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{% load i18n %}
+{% include "wagtailadmin/shared/header.html" with title=action_verbose icon="clipboard-list" %}
+
+<div class="nice-padding">
+    <form action="{% url 'wagtailadmin_pages:workflow_action' page.id action task_state.id %}" method="POST" novalidate>
+        {% csrf_token %}
+        <ul class="fields">
+            {{ form.as_ul }}
```
[the remark this answers is not in the document]
Comment (jacobtoppm) [reply wagtail/admin/templates/wagtailadmin/pages/workflow_action_modal.html]: Thanks, I hadn't seen that! I've created the include, but haven't refactored other uses yet (as I want to be able to test that everything's still working more thoroughly after I do)

File: wagtail/admin/static_src/wagtailadmin/js/workflow-action.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                                            // if the response can be read as JSON, update the modal with its contents
+                                            // this is intended for form errors
+                                            modal.loadResponseText(responseText);
+                                        } 
+                                        catch(err) {
+                                            // the response is not JSON, and not intended for the modal - the action succeeded
+                                            // show the response content instead of the page
+                                            document.open("text/html", "replace");
+                                            document.write(responseText);
+                                            document.close();
```
[the remark this answers is not in the document]
Comment (jacobtoppm) [reply wagtail/admin/static_src/wagtailadmin/js/workflow-action.js]: Hah, I agree, I was dubious about this too, but had a bit of a mind blank on how to do it more neatly. You're right, that's a lot neater, thanks. Reworked now!

File: wagtail/admin/views/pages.py [hunk 5]
```diff
@@ -1279,25 +1286,63 @@ def workflow_action(request, page_id):
         messages.error(request, _("The page '{0}' is not currently awaiting moderation.").format(page.get_admin_display_title()))
         return redirect(redirect_to)
 
-    task_state_id = request.POST.get('task_state_id', None)
-
-    task_state = TaskState.objects.get(id=task_state_id) if task_state_id else page.current_workflow_task_state
+    task_state = get_object_or_404(TaskState, id=task_state_id) if task_state_id else page.current_workflow_task_state
```
[the remark this answers is not in the document]
Comment (jacobtoppm) [reply wagtail/admin/views/pages.py]: Thanks, fixed now!

File: wagtail/admin/mail.py (hunk 1 above)
Comment (jacobtoppm) [reply wagtail/admin/mail.py]: I agree, I couldn't think of a neat way to do this either. I've reverted for now - it may not be necessary to have this.

**Pull request merged**